### PR TITLE
Generate new UUIDs web when duplicating jobs

### LINKED ISSUE #1802: [Submission detail] "No data collected" message unclear
Should say "No data collected for this site"

image

@amegantz FYI

### LINKED ISSUE #1803: Missing duplicated jobs in Android
Cause: duplicate job doesn't generates different task ids

### PR DESCRIPTION
**Pull request by rfontanarosa**
closes #1802
closes #1803

Screenshot 2024-05-30 alle 15 26 23

### CODE REVIEW

File: web/src/app/components/tasks-editor/tasks-editor.component.ts [hunk 1]
```diff
@@ -176,23 +178,7 @@ export class TasksEditorComponent {
...
-                  ...task.multipleChoice,
-                  options: task.multipleChoice.options?.map(option => ({
-                    ...option,
-                    id: this.dataStoreService.generateId(),
-                  })),
-                } as MultipleChoice)
-              : undefined,
-            task.condition
-          );
+          const duplicatedTask = this.taskService.createDuplicateTask(task);
```
Comment (gino-m) [inline web/src/app/components/tasks-editor/tasks-editor.component.ts:181]: Nit: Why not call this method just `duplicateTask(task)`?

File: web/src/app/pages/edit-survey/edit-survey.component.spec.ts [hunk 2]
```diff
@@ -262,11 +263,7 @@ describe('EditSurveyComponent', () => {
         duplicateButton.click();
 
         expect(draftSurveyServiceSpy.addOrUpdateJob).toHaveBeenCalledOnceWith(
-          job1.copyWith({
-            id: newJob.id,
-            name: `Copy of ${job1.name}`,
-            color: undefined,
-          }),
+          jobServiceSpy.createDuplicatedJob(job1, survey.jobs),
```
Comment (gino-m) [inline web/src/app/pages/edit-survey/edit-survey.component.spec.ts:266]: Same here: `duplicateJob()`?

File: web/src/app/pages/edit-survey/edit-survey.component.ts [hunk 3]
```diff
@@ -111,21 +111,16 @@ export class EditSurveyComponent implements OnInit {
...
-
-    this.draftSurveyService.addOrUpdateJob(
-      job.copyWith({
-        id: newJob.id,
-        name: `Copy of ${job.name}`,
-        color: this.jobService.getNextColor(this.survey?.jobs),
-        index: -1,
-      }),
-      true
+    const duplicatedJob = this.jobService.createDuplicatedJob(
```
Comment (gino-m) [inline web/src/app/pages/edit-survey/edit-survey.component.ts:114]: `duplicatedJob` sounds a bit odd - `newJob` works better here imo.

File: web/src/app/services/job/job.service.ts [hunk 4]
```diff
@@ -70,6 +73,24 @@ export class JobService {
     );
   }
 
+  /**
+   * Returns a new job which is an exact copy of the specified job, but with new UUIDs for all items recursively.
+   */
+  createDuplicatedJob(job: Job, jobs: Map<string, Job> | undefined): Job {
```
Comment (gino-m) [inline web/src/app/services/job/job.service.ts:79]: passing `job` and `jobs` seems redundant - maybe just pass `Survey` and the id of the job to duplicate?

File: web/src/app/services/task/task.service.ts [hunk 5]
```diff
@@ -69,6 +69,25 @@ export class TaskService {
...
+   * Returns a new task which is an exact copy of the provided task, but with new UUIDs for all items recursively.
+   */
+  createDuplicateTask(task: Task): Task {
+    return {
+      ...task,
+      id: this.dataStoreService.generateId(),
+      multipleChoice: task.multipleChoice
+        ? ({
+            ...task.multipleChoice,
+            options: task.multipleChoice.options?.map(option => ({
```
Comment (gino-m) [inline web/src/app/services/task/task.service.ts:82]: Nit: For consistency and readability, you could define this in its own method, `duplicateMultipleChoiceOption`?

**Review by gino-m — CHANGES_REQUESTED** [review]

File: web/src/app/services/job/job.service.ts (hunk 4 above)
Comment (rfontanarosa) [reply web/src/app/services/job/job.service.ts:79]: I've solved passing the color as already done with createNewJob.

**Review by gino-m — APPROVED** [review]
✌️

**Pull request merged**